### PR TITLE
[c10d] Make send/recv as custom ops

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79779

Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

### CODE REVIEW

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 1]
```diff
@@ -118,6 +118,26 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
       BarrierOptions{device_ids, std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> send(
+    at::TensorList tensors,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:122]: same comment on `TensorList` vs `std::vector<Tensor>`

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 2]
```diff
@@ -118,6 +118,26 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
       BarrierOptions{device_ids, std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> send(
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:121]: trailing underscore?

Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:121]: Send is not a in-place op. So there is no needs to have a trailing _.

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:122]: I think it's a convention to use TensorList to represent const std::vectorat::Tensor&.

I only use const std::vectorat::Tensor& if there is a const std::vector<std::vectorat::Tensor>& in the signature to keep consistency on that function. Let me know which way you like.

**Review by mrshenli — APPROVED** [review]

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 3]
```diff
@@ -328,5 +350,35 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
   return op.call(process_group, opts.device_ids, opts.timeout.count());
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> send(
+    const c10::intrusive_ptr<ProcessGroup>& process_group,
+    at::TensorList tensors,
+    int64_t dstRank,
+    int64_t tag) {
+  static auto op = c10::Dispatcher::singleton()
+                       .findSchemaOrThrow("c10d::send", "")
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/Ops.cpp:359]: hmm would we not find the schema by any chance?

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 4]
```diff
@@ -328,5 +350,35 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
...
+                       .typed<c10::intrusive_ptr<::c10d::ProcessGroup::Work>(
+                           at::TensorList,
+                           const c10::intrusive_ptr<::c10d::ProcessGroup>&,
+                           int64_t,
+                           int64_t)>();
+  return op.call(tensors, process_group, dstRank, tag);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> recv(
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/Ops.cpp:368]: just noticed we have this `recv` and `recv_` two separate APIs (above), what's the difference between these two and should we merge?

**Review by wanchaol — APPROVED** [review]
lgtm, have some questions.

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 3 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:359]: Why would you have such concern? Do you happen to know another op called "c10d::send" as well?

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 4 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:368]: The above recv_ is the implementation of the function schema. This one is the C++ API for users to call the op. recv_ follows aten ops's conventions while recv follows c10d's conventions.

**Comment by alanwaketan** [comment]
Thanks for approving this pull request, Shen and Wanchao.

**Pull request closed without merging**